v2/src/main/scala/coop/rchain/v2/casper/data/CasperScope.scala

```diff
@@ -0,0 +1,17 @@
+package coop.rchain.v2.casper.data
+
+/**
+ * Scope containing all information required for Casper to make decisions: finalized state of the network and all
+ * messages that are not finalized.
+ *
+ * Casper scope is valid only if finalization fringe is complete: finalized messages are found for all known senders.
+ *
+ * @param latestMessages      Upper boundary, latest messages known. Define the whole [[CasperScope]].
+ * @param finalizationFringe  Bottom boundary, set of finalized messages for all known senders.
+ * @param conflictScope       The body of the scope - all messages above [[finalizationFringe]].
+ */
+final case class CasperScope[M, S](
+    latestMessages: LatestMessages[M, S],
+    finalizationFringe: FinalizationFringe[M, S],
+    conflictScope: ConflictScope[M]
+) { require(finalizationFringe.v.keySet == latestMessages.v.keySet) }
```

Comment:
`Casper scope is valid only if finalization fringe is complete: finalized messages are found for all known senders.`

What does `valid` mean in the above sentence?

Can I say "For a particular block, there is always a fixed value `CasperScope` corresponding to this particular block" is a correct property for a block?

Reply:
Yes, you are right, for each block there is Casper scope associated. When you replay - you reconstruct the scope. To be able to create a block, proposer first have to find the scope, as well.

`Valid` means that if there is no finalized message for all senders in a fringe - Casper scope cannot be created. So if you will try to create such an instance - you will hit an error. I need to add here error msg actually.